clippy_lints/src/loops/mut_range_bound.rs

```diff
@@ -116,3 +120,49 @@ impl MutatePairDelegate<'_, '_> {
         (self.span_low, self.span_high)
     }
 }
+
+struct BreakAfterExprVisitor {
+    hir_id: HirId,
+    past_expr: bool,
+    break_after_expr: bool,
+}
+
+impl BreakAfterExprVisitor {
+    pub fn is_found(cx: &LateContext<'_>, hir_id: HirId) -> bool {
+        let mut visitor = BreakAfterExprVisitor {
+            hir_id,
+            past_expr: false,
+            break_after_expr: false,
+        };
+
+        get_enclosing_block(cx, hir_id).map_or(false, |block| {
+            visitor.visit_block(block);
+            visitor.break_after_expr
+        })
+    }
+}
+
+impl intravisit::Visitor<'tcx> for BreakAfterExprVisitor {
+    type Map = Map<'tcx>;
+
+    fn nested_visit_map(&mut self) -> NestedVisitorMap<Self::Map> {
+        NestedVisitorMap::None
+    }
+
+    fn visit_expr(&mut self, expr: &'tcx Expr<'tcx>) {
+        if self.break_after_expr {
+            return;
+        }
+        if matches!(&expr.kind, ExprKind::If(..)) || matches!(&expr.kind, ExprKind::Match(..)) {
+            return;
+        }
```

Comment:
I think this is there to make sure that control flow doesn't get in the way? There are way more things that could break this analysis, like other loops for example.

I think you want to achieve, that `break;` doesn't have to follow the assignment immediately. 

I wouldn't care about this that much, since you should be able to move the assignment immediately in front of the `break`. E.g.:

```rs
m = some_expr;
x = 2 + m;
break;
```

could also be written as

```rs
res = some_expr
x = 2 + res;
m = res;
break;
```

---

On the other hand, it would be nice, if this would also work on the above example, but this is difficult to implement correctly.

Reply:
I'm not sure that I completely follow the example here

Reply:
I'm not sure if I understand the code in the visitor correctly. Why do you check for `if` and `match`?

Reply:
This is to cover for

https://github.com/dswij/rust-clippy/commit/de0f3de605e6d4c51c812322692df545a6cad343#diff-26bc8b24b8058cb2b9f8fdc7ed87d6cb272e71386db63e328905d827ec242adaR89-R99

The `break` inside the control flow is not always reachable, so I think it should still be warned. The check is there to ignore `break`s inside control flow. 

That being said, I do agree that there are more things that might break this, such as #6072 😕

Reply:
Yes, I thought this would be the reason. 

So I think this would be easier to implement, to just allow `break;`s  right after the expression that mutates the range bound variable. This would mean that

```rs
m = 5;
some_other_expr;
break;
```

would still lint, but 

```rs
some_other_expr;
m = 5;
break;
```

would no longer lint.

Reply:
Ah OK, I understand the first comment now. 

Yes, I think that this actually makes more sense in term of what we want to achieve rather than complicate things.